tpu-client/src/quic_client.rs

```diff
@@ -15,11 +15,59 @@ use {
     },
     lazy_static::lazy_static,
     solana_sdk::transport::Result as TransportResult,
-    std::{net::SocketAddr, sync::Arc},
+    std::{
+        net::SocketAddr,
+        sync::{Arc, Condvar, Mutex, MutexGuard},
+    },
     tokio::runtime::Runtime,
 };
 
+const MAX_OUTSTANDING_TASK: u64 = 2000;
+
+/// A semaphore used for limiting the number of asynchronous tasks spawn to the
+/// runtime. Before spawnning a task, use acquire. After the task is done (be it
+/// succsess or failure), call release.
+struct AsyncTaskSemaphore {
```

Comment:
Any benefits of this over https://docs.rs/tokio/0.2.6/tokio/sync/struct.Semaphore.html?

Reply:
Yes, I looked at the Tokio implementation. The acquire interface is an async function. We need a sync function here before spawning.

Reply:
Wait, nevermind, read that wrong. So we're actually blocking before spawning the async task... would this potentially block the spawning thread unduly? We did originally roll the whole async spawn thing because block_on was blocking the spawning task for too long...

Reply:
This only blocks during the short time of executing the spawn itself -- not during exciting actual async function. The spawn function itself is acquiring an internal mutex actually to protect the structures. The internal lock is also acquired when the task is removed -- after finishing. I do not expect this cause any long block of the runtime itself when executing tasks.

Reply:
Ok, fair enough. Are there any pre-built semaphore primitives we could use instead though? I really don't like rolling our own with these kinds of things. If not, that's fine...

Reply:
I have not found any other. It used to had std::sync::Semaphore -- it was deprecated and removed. As I said Tokio one is for running inside async functions only not for this purpose. The logic of using Cond variable with a mutex is widely used pattern in both C and rust -- I do not see this anything out of extraordinary.

Reply:
Fair enough, I don't see anything wrong with the implementation.

Reply:
always feels weird to see a implemented built from things that are almost certainly built around semaphores under the hood, but yes indeed there's no good implementation in rust.  closest i've seen is the posix ones out of `libc`

Reply:
@lijunwangs I read through this and I still don't understand why we are using a blocking semaphore here.
Can you provide more details why an async semaphore wouldn't work here? Like in what ways an async semaphore would allow more tasks to be spawned compared to the blocking one here?

To my understanding the blocking semaphore will prevent other already-spawned tasks to execute in place of the blocked one; which is not desirable.

Reply:
As explained in the problem statement of this PR, the analysis was we have spawned too many asynchronous tasks doing the sends into the runtime that results in timeout in streams (streams not completed fast enough in time) and connections (connection heartbeat for quic not being served fast enough as it is also driven by the runtime). So the idea was to prevent too many async task being spawned into the Runtime. Note before spawn it is a sync call context, hence the need a synch lock. The lock is released after the task is finished (either success or failure) by the runtime.

Reply:
FYI, I am revising this design in the context of repair using Quic. I was concerned too many connections to problematic endpoint clogging this and cause problems for other legitimate endpoints -- which can be attack vector. One could just send requests without serving the responses. So the one Semaphore for all connections can be problematic here. We need to a better per endpoint limit mechanism. @behzadnouri

Reply:
In addition -- I did some testing on this limit -- the current master code with the latest Quinn code seems to be much more tolerant of the higher value than before.

Reply:
i think we should just dump the excess connection attempts on the floor. don't even bother spawning a task or evaluating them.

Reply:
> The lock is released after the task is finished (either success or failure) by the runtime.

@lijunwangs But the lock is released inside the task before the task is done:
https://github.com/solana-labs/solana/blob/fdecd0dca/quic-client/src/quic_client.rs#L83
https://github.com/solana-labs/solana/blob/fdecd0dca/quic-client/src/quic_client.rs#L98

isn't this wrong?

Reply:
The task is essentially done -- note the function calls remaining are the unlock and update stats which are all sync code . The worker thread from the runtime will not be switched to run other async tasks during this.

Reply:
> i think we should just dump the excess connection attempts on the floor. don't even bother spawning a task or evaluating them.

I was thinking of the same -- but we do need a counter per destination address to track it.

Reply:
> I was thinking of the same -- but we do need a counter per destination address to track it.

well if we're in a situation where we need to load shed, i'm not sure we should bother address tracking.  similarly, we could probably just use a `Deduper` and accept some false positives